### PR TITLE
Fix runtime panic with L2 announcer name generation

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Since we upgrade the `hive` in [this PR](https://github.com/cilium/cilium/pull/34920), ConformanceGatewayAPI is failing consistently because the L2 announcer is hitting the newly introduced Job name limit.

```
2024-09-24T19:58:20.700855025Z panic: invalid job name: "leader-election-gateway-conformance-infra-cilium-gateway-http-listener-isolation-with-hostname-h56mb54hf8", expected to match "^[a-z][a-z0-9_\\-]{0,100}$"
2024-09-24T19:58:20.700873208Z 
2024-09-24T19:58:20.700876094Z goroutine 356 [running]:
2024-09-24T19:58:20.700878689Z github.com/cilium/hive/job.OneShot({0xc000ded1f0, 0x69}, 0xc0008cb470, {0x0, 0x0, 0x0})
2024-09-24T19:58:20.700881274Z 	/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/oneshot.go:27 +0xe5
2024-09-24T19:58:20.700892354Z github.com/cilium/cilium/pkg/l2announcer.(*L2Announcer).addSelectedService(0xc00208a240, 0xc002a5f0e0, {0xc0015486e0, 0x1, 0x1})
2024-09-24T19:58:20.700898256Z 	/go/src/github.com/cilium/cilium/pkg/l2announcer/l2announcer.go:742 +0x29f
2024-09-24T19:58:20.700919395Z github.com/cilium/cilium/pkg/l2announcer.(*L2Announcer).upsertPolicy(0xc00208a240, {0x4a2a990, 0xc0007a9ad0}, 0xc0027ae2c0)
2024-09-24T19:58:20.700922020Z 	/go/src/github.com/cilium/cilium/pkg/l2announcer/l2announcer.go:559 +0xe65
2024-09-24T19:58:20.700928001Z github.com/cilium/cilium/pkg/l2announcer.(*L2Announcer).upsertLocalNode(0xc00208a240, {0x4a2a990, 0xc0007a9ad0}, 0xc0008e69f0?)
2024-09-24T19:58:20.700931167Z 	/go/src/github.com/cilium/cilium/pkg/l2announcer/l2announcer.go:859 +0x4d7
2024-09-24T19:58:20.700940654Z github.com/cilium/cilium/pkg/l2announcer.(*L2Announcer).processLocalNodeEvent(0x0?, {0x4a2a990?, 0xc0007a9ad0?}, {{0x435984f, 0x6}, {{0xc0028ca330, 0x14}, {0x0, 0x0}}, 0xc00123f808, ...})
2024-09-24T19:58:20.700955522Z 	/go/src/github.com/cilium/cilium/pkg/l2announcer/l2announcer.go:808 +0x52
2024-09-24T19:58:20.700963447Z github.com/cilium/cilium/pkg/l2announcer.(*L2Announcer).run(0xc00208a240, {0x4a2a990, 0xc0007a9ad0}, {0xc002b0fd78?, 0x4afcaf?})
2024-09-24T19:58:20.700970981Z 	/go/src/github.com/cilium/cilium/pkg/l2announcer/l2announcer.go:164 +0x41d
2024-09-24T19:58:20.701000025Z github.com/cilium/hive/job.(*jobOneShot).start(0xc001e28cc0, {0x4a2a990, 0xc0007a9ad0}, 0x0?, {0x4a33960, 0xc001e28c00}, {{{0xc000df99e0, 0x1, 0x1}}, 0xc000edc5d0, ...})
2024-09-24T19:58:20.701005265Z 	/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/oneshot.go:140 +0x503
2024-09-24T19:58:20.701012609Z created by github.com/cilium/hive/job.(*group).Start.func1 in goroutine 1
2024-09-24T19:58:20.701019973Z 	/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/job.go:161 +0x174
```

Fixes: #34920 

```release-note
Fix runtime panic with L2announcer name generation
```

### CODE REVIEW

**Review by youngnick — APPROVED** [review]

File: pkg/shortener/shortener.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"crypto/sha256"
+	"fmt"
+)
+
+const (
+	// Maximum characters in a K8s resource name
+	k8sMaxResourceNameLength = 63
+)
+
+func ShortenK8sResourceName(s string) string {
```
Comment (bimmlerd) [inline pkg/shortener/shortener.go:19]: I don't want to block the PR on it, but could use a bit of unit testing to document/encode that a property this needs to have is that it's pure/stable in the sense that it must return the same short name for a given input.

**Review by bimmlerd — APPROVED** [review]
Did you think about whether this can have some sort of upgrade impact? I could not think of anything, but in the general direction of some subsystem not finding some persistent state due to the name being now shortened?

**Review by sayboras — APPROVED** [review]
Thanks, and looks good for envoy related file.

File: pkg/shortener/shortener.go (hunk 1 above)
Comment (YutaroHayakawa) [reply pkg/shortener/shortener.go:19]: Yep, makes sense. I'll add that.

**Comment by YutaroHayakawa** [comment]
For the upgrade impact, I couldn't come up with anything. At a glance there's no code that refer this job by name. One thing I can come up is a metrics name. If someone has for example a PromQL that is aware of this `leader-election-<svc namespace>-<svc name>` structure, this may break it. However, IMO they shouldn't rely on it because it's an internal implementation details.

File: pkg/shortener/shortener.go (hunk 1 above)
Comment (YutaroHayakawa) [reply pkg/shortener/shortener.go:19]: Added

**Comment by YutaroHayakawa** [comment]
Conformance Runtime: https://github.com/cilium/cilium/issues/34557

**Pull request merged**